src/util/redux.spec.js

```diff
@@ -0,0 +1,197 @@
+
+	const reducers = {
+		foo: {
+			onChange: (state, payload) => {
+				return ({ value: payload });
+			},
+			bar: {
+				onChange: (state, payload) => ({ value: payload })
+			},
+			asyncOperation: thunk(payload => dispatchTree => Promise.resolve(payload).then(dispatchTree.onChange))
+		}
+	};
+
+	const initialState = {
+		foo: {
+			value: 'foo',
+			bar: {
+				value: null
+			}
+		}
+	};
+
+	describe('reducer', () => {
+
+		const { reducer } = getReduxPrimitives({ reducers, initialState });
+
+		it('should return initialState on unmatched type', () => {
+			const action = { type: 'UNKNOWN' };
+			assert.deepEqual(reducer(initialState, action), initialState);
+		});
+
+		it('should correctly apply state change', () => {
+			const action = { type: 'foo.onChange', payload: 'bar' };
+			const nextState = reducer(initialState, action);
+			assert.equal(nextState.foo.value, 'bar');
+		});
+
+		describe('nested reducer', () => {
+
+			it('should correctly apply state change', () => {
+				const action = { type: 'foo.bar.onChange', payload: 'baz' };
+				const nextState = reducer(initialState, action);
+				assert.equal(nextState.foo.bar.value, 'baz');
+			});
+
+		});
+
+		describe('with rootPath', () => {
+
+			const { reducer: reducerWithRootPath } = getReduxPrimitives({
+				reducers,
+				initialState,
+				rootPath: ['root']
+			});
+
+			it('should correctly apply state change', () => {
+				const action = { type: 'root.foo.onChange', payload: 'bar' };
+				const nextState = reducerWithRootPath(initialState, action);
+				assert.equal(nextState.foo.value, 'bar');
+			});
+
+		});
+
+		describe('thunks', () => {
+			it('should not include thunk paths in reducer', () => {
+				const action = { type: 'root.foo.asyncOperation' };
+				const nextState = reducer(initialState, action);
+				assert.deepEqual(initialState, nextState);
+			});
+		});
+
+	});
+
+	describe('connectors', () => {
+		describe('selector/mapStateToProps', () => {
+
+			const selectors = {
+				foo: {
+					uppercase: ({ value }) => upperCase(value)
+				}
+			};
+
+			const {
+				connectors: [mapStateToProps]
+			} = getReduxPrimitives({
+				reducers,
+				initialState,
+				selectors
+			});
+
+			it('should apply selector', () => {
+				const viewState = mapStateToProps(initialState);
+				assert.equal(viewState.foo.uppercase, 'FOO');
+			});
+
+			describe('rootPath', () => {
+
+				const {
+					connectors: [mapStateToProps]
+				} = getReduxPrimitives({
+					reducers,
+					initialState,
+					selectors,
+					rootPath: ['root']
+				});
+
+				it('should apply selector', () => {
+					const viewState = mapStateToProps({ root: initialState });
+					assert.equal(viewState.foo.uppercase, 'FOO');
+				});
+
+			});
+
+			describe('rootSelector', () => {
+
+				const {
+					connectors: [mapStateToProps]
+				} = getReduxPrimitives({
+					reducers,
+					initialState,
+					selectors,
+					rootSelector: state => assign({}, state, {
```

Comment:
Do you envision this extra layer of computed values to be useful for things like validation? I like the approach.